woocommerce/woocommerce-gateway-stripe
# SCA Subs: Allow SCA challenges in the early renewals modal

Fixes #1049.

#### Changes proposed in this Pull Request:

The Early Renewals Modal of WooCommerce Subscriptions contains a single button (link), which triggers the payment of the next renewal order for the subscription.

In order to make the link SCA-compatible:

1. There is a new handler for the link in `stripe.js`: `onEarlyRenewalSubmit`.
2. In order to keep `onEarlyRenewalSubmit` on working even when no SCA is required, the `redirect_after_early_renewal` method of the compatibility class makes sure that the value, which is sent to JS is always an array, which contains the URL to redirect to.
3. When the `process_early_renewal` flag is present in the request, `process_subscription_renewal` forwards the action back to the normal gateway's `process_payment` method, since the payment is not an off-session one.
4. A couple of actions within the intent verification method allow the compatibility class to perform the necessary actions for early renewals:
    - Reset the subscription's dates when the renewal was completed successfully, even with an SCA challenge.
    - Delete the renewal order and redirect to the checkout page for normal checkout, when the SCA challenge is failed.

#### Testing instructions

1. Setup Subscriptions in order to see the "Renew Now" button on a subscription's page.
2. Make sure the subscription uses a non-SCA card (4242) and renew. It should go smoothly.
3. Switch to the SCA card, which ends with 3184 and always requires SCA challenges.
4. Try to renew the order (without intentionally failing). It should go smoothly, with a notification of what happened.
5. Try to renew again, but fail the SCA challenge. You should be redirected to the checkout with a message, which says "Payment authorization for the renewal order was unsuccessful, please try again.".

-------------------
